### PR TITLE
fix: preserve crawler data integrity

### PR DESCRIPTION
**Pull request by hiroppy**
# Summary

- Make crawler save and group restoration atomic so incomplete crawls do not publish partial data or report success.
- Wait for Money Forward cash-flow transitions and validate every rendered transaction row before replacing a period.
- Preserve custom cross-month accounting periods, year boundaries, empty valid descriptions, and end-day timestamps.
- Keep category decisions on the authenticated history snapshot instead of re-scraping through a range URL that the service ignores.
- Reject missing IDs, unsupported amounts, and failed required-cell extraction before database replacement.
- Keep crawler diagnostics anonymous.

This contains the production bug fixes discovered while making crawler E2E strict. PR #310 will be rebased conceptually by changing its base to this branch, leaving #310 with only E2E/CI changes.

## Linear Issue

- N/A

### Selected Quality Characteristics

| Quality characteristic | Why it is relevant | Acceptance criteria |
| ---------------------- | ------------------ | ------------------- |
| Functional suitability | Transactions, amounts, transfer relationships, and accounting periods must match the authenticated service | Every extracted period is saved with the same count; common transaction IDs retain identical amounts, types, transfer flags, and account links |
| Reliability | Partial DOM reads, failed saves, and failed state restoration must not publish inconsistent results | Incomplete extraction aborts replacement; database writes roll back atomically; completion is reported only after restoration |
| Information security | Real-account validation must not expose personal financial data | Tests and PR evidence use anonymous fixtures and aggregate comparison counts only |

### Test Techniques

| Test technique | Selection rationale | Expected coverage |
| -------------- | ------------------- | ----------------- |
| Boundary value analysis | Month/year boundaries and end-date timestamps were high-risk | Cross-month and cross-year dates, inclusive end day, and exclusive next day |
| Decision table testing | Empty values, missing IDs, unsupported amounts, and extraction failures require distinct outcomes | Valid empty descriptions succeed; invalid or failed required fields abort |
| State transition testing | Refresh, save, group restoration, and completion have ordered states | Failures roll back and prevent false success; successful runs restore state before completion |
| Error guessing | AJAX replacement and stale DOM behavior caused previously hidden corruption | Table replacement, unchanged periods, missing rows, and ignored range navigation |
| Exploratory/checklist testing | Production-like integrity needs reconciliation against the authenticated service | Fresh DB integrity and aggregate comparison with the existing DB |

### Primary Test Conditions

- Complete and incomplete transaction rows, including empty valid descriptions.
- Current month, custom cross-month periods, and cross-year periods.
- End-date timestamps and out-of-range records.
- Full and partial category-update success.
- Save failures and group-restoration failures.
- Fresh authenticated history crawl with per-period extraction/save reconciliation.
- Transfer amount, type, flag, source account, and target account reconciliation.

### Out-of-Scope Risks

- Current account totals and holding valuations can legitimately change between crawls; exact equality is not required.
- The older DB was not mutated during comparison.

## Verification

- [x] Unit tests
- [x] Type checking
- [x] Lint
- [x] Storybook tests
- [x] E2E tests
- [x] Manual verification

### Commands Executed

```text
pnpm test — passed
pnpm turbo typecheck — passed (8/8 packages)
pnpm lint — passed
pnpm format:check — passed
pnpm knip — passed (existing configuration hints only)
pnpm --filter @mf-dashboard/crawler test:e2e — passed (74/74, skipped 0)
DB_PATH=<temporary> SCRAPE_MODE=history CLEANUP_GROUPS=false pnpm --filter @mf-dashboard/crawler dev:scrape — passed
fresh crawl reconciliation — 4,957/4,957 extracted/saved, 20/20 periods, per-period mismatches 0
existing DB reconciliation — 4,942 common IDs; amount/type/transfer/source/target mismatches 0
SQLite integrity and transaction invariant checks — passed
```

### Checks Not Run

- N/A — all relevant checks were run.

## Summary by CodeRabbit

* **Bug Fixes**
  * Improved cash-flow scraping across month boundaries and calendar years.
  * Added validation for invalid, incomplete, or unsupported transaction data.
  * Ensured empty monthly results correctly remove outdated transactions.
  * Improved recovery when switching groups or restoring crawler state.
  * Categorization now preserves supplied cash-flow data and applies partial decisions safely.

* **Improvements**
  * Historical transactions now save using precise date ranges and completion status.
  * Current and previous periods refresh reliably.
  * Logs no longer expose private group names, identifiers, or balance details.
  * Notifications and refreshes are suppressed when restoration fails.
  * Added tracking for downloaded cash-flow periods and their completeness.

### CODE REVIEW

**Comment by hiroppy** [comment]
closes: https://github.com/hiroppy/mf-dashboard/issues/315

File: apps/crawler/src/scrapers/cash-flow-history.ts [hunk 1]
```diff
@@ -139,31 +266,48 @@ async function detectTransactionType(
...
+    /^\d{4}-\d{2}-\d{2}$/.test(date) &&
+    !Number.isNaN(parsedDate.getTime()) &&
+    parsedDate.toISOString().slice(0, 10) === date;
+
+  // A monthly replacement is safe only when every rendered transaction row was extracted.
+  const incompleteFields = [
+    !mfId ? "id" : null,
+    rowClass === undefined ? "class" : null,
+    !isValidDate ? "date" : null,
+    description === null ? "description" : null,
```
[the remark this answers is not in the document]
Comment (hiroppy) [reply apps/crawler/src/scrapers/cash-flow-history.ts:319]: 対応しました。E2Eは description の型だけを検証し、正常に抽出された空文字を許可します。全E2E 74/74、skip 0で確認済みです。

File: apps/crawler/src/scrapers/cash-flow.ts [hunk 2]
```diff
@@ -4,7 +4,55 @@ import { mfUrls } from "@mf-dashboard/meta/urls";
...
+  SUMMARY_COLUMNS,
+  parseDetailRow,
+  resolveCashFlowPeriod,
+  waitForCashFlowFetchApplied,
+} from "./cash-flow-history.js";
+
+export function parseCashFlowMonthHeader(headerText: string | null): string | null {
+  const match =
+    headerText?.match(/(\d{4})年(\d{1,2})月/) ??
+    headerText?.match(/(\d{4})\/(\d{1,2})\/\d{1,2}\s*-/);
```
[the remark this answers is not in the document]
Comment (hiroppy) [reply apps/crawler/src/scrapers/cash-flow.ts]: 対応しました。月跨ぎヘッダーは期間終了月から判定する共通パーサーへ集約し、CSVリンクなし分岐を認証済みread-only E2Eで検証しています。commit: 6de3a4e（後続simplify含む）。全E2E 74/74です。

File: apps/crawler/src/category-decision/categorize-cash-flow.ts [hunk 3]
```diff
@@ -93,25 +92,15 @@ export async function categorizeCashFlowMonth(options: {
...
-    }
-
-    const latestCashFlow = await scrapeCashFlowMonth(page, cashFlow.month);
-    latestCashFlowForFallback = latestCashFlow;
-    const latestTargets = await findCategorizationTargets(db, latestCashFlow);
-    if (latestTargets.length === 0) {
-      return latestCashFlow;
-    }
+    const targets = await findCategorizationTargets(db, cashFlow);
+    if (targets.length === 0) return cashFlow;
```
[the remark this answers is not in the document]
Comment (hiroppy) [reply apps/crawler/src/category-decision/categorize-cash-flow.ts:98]: 対応しました。履歴結果がある場合は runSavePhase 側の当月カテゴリ判定をスキップし、履歴側で分類済みの結果をatomic保存するため、LLM/API更新は一度だけです。unit testで二重呼び出しがないことを確認しました。

File: apps/crawler/src/crawler-phases.ts [hunk 4]
```diff
@@ -358,14 +363,18 @@ export async function runCashFlowHistoryPhase(
           })
         : monthData;
       preparedMonths.push({
+        dateRange:
+          categorizedMonthData.periodStart && categorizedMonthData.periodEnd
+            ? { from: categorizedMonthData.periodStart, to: categorizedMonthData.periodEnd }
+            : undefined,
```
[the remark this answers is not in the document]
Comment (hiroppy) [reply apps/crawler/src/crawler-phases.ts:363]: 対応しました。取引日ではなく cash_flow_periods に保存済み会計期間を記録し、同一transaction内で期間カバレッジも更新します。migrationとDB設計書も追加済みです。DB 395件、typecheck 8/8で確認しました。

File: packages/db/src/repositories/save-scraped-data.ts [hunk 5]
```diff
@@ -306,13 +310,15 @@ async function saveScrapedDataAtomically(db: DbExecutor, data: ScrapedData): Pro
...
-    await saveTransaction(db, item, accountIdMap);
-    if (item.mfId && !item.mfId.startsWith("unknown")) {
-      savedCount++;
-    }
-  }
+  const savedCount = await replaceTransactionsForMonth(
+    db,
+    data.cashFlow.month,
+    data.cashFlow.items,
+    accountIdMap,
```
[the remark this answers is not in the document]
Comment (hiroppy) [reply packages/db/src/repositories/save-scraped-data.ts:319]: 対応しました。スクレーパーが必須サマリーを正常取得した場合だけ isComplete を付与し、空期間は completeness proof がなければ置換を拒否します。無効IDも黙って除外せず削除前に失敗させます。DB unit 395件で確認済みです。

File: apps/crawler/src/scrapers/cash-flow-history.ts [hunk 6]
```diff
@@ -309,10 +441,13 @@ export async function scrapeCashFlowHistory(
...
-        await Promise.all([
-          page.waitForResponse((res) => res.url().includes("/cf/fetch") && res.status() === 200),
-          prevButton.click(),
-        ]);
+        await waitForCashFlowFetchApplied(page, async () => {
+          const [fetchResponse] = await Promise.all([
+            page.waitForResponse((res) => res.url().includes("/cf/fetch") && res.status() === 200),
+            prevButton.click(),
+          ]);
+          await fetchResponse.finished();
```
[the remark this answers is not in the document]
Comment (hiroppy) [reply apps/crawler/src/scrapers/cash-flow-history.ts]: 対応しました。Response.finished() の戻り値を検査し、本文受信エラーなら履歴末尾として扱わずcrawlを失敗させます。該当failure branchのunit testを追加済みです。

File: apps/crawler/src/scraper-progress.test.ts [hunk 7]
```diff
@@ -90,6 +90,42 @@ afterEach(async () => {
...
+      });
+      const diagnostics = JSON.stringify(consoleLog.mock.calls);
+      expect(diagnostics).not.toContain("private-group-id");
+      expect(diagnostics).not.toContain("private-group-name");
+      expect(diagnostics).not.toContain("private-balance");
+    } finally {
+      consoleLog.mockRestore();
+    }
+  });
+
```
[the remark this answers is not in the document]
Comment (hiroppy) [reply apps/crawler/src/scraper-progress.test.ts:152]: 対応しました。logger をモックして CI 設定に依存しないようにし、console.log/warn/error と logger の各出力をまとめて匿名性を検証しています。commit: 6de3a4e。crawler unit 352件、全E2E 74件（skip 0）で確認済みです。

File: apps/crawler/src/crawler-phases.ts [hunk 8]
```diff
@@ -257,28 +257,31 @@ export async function runCashFlowHistoryPhase(
...
 
-  let monthsToFetch = 1;
-  for (let i = 1; i < maxMonths; i++) {
-    const month = getHistoryMonth(now, i);
-    if (!(await hasTransactionsForMonth(db, month))) {
-      monthsToFetch = i + 1;
+  // Always refresh the current and previous periods so transactions posted late by an
+  // institution are incorporated. History mode extends that window to the oldest gap.
+  let monthsToFetch = Math.min(2, maxMonths);
+  if (config.isHistoryMode) {
```
[the remark this answers is not in the document]
Comment (hiroppy) [reply apps/crawler/src/crawler-phases.ts:279]: 対応しました。month mode で当月＋直前会計期間を再取得するのは、遅延反映された取引とカテゴリ更新を取り込むための意図した仕様です。AGENTS.md の旧記述を現在の契約へ更新し、history mode の範囲も実装どおり明記しました。直前期間の取得に失敗した場合は save phase 前に中断するため、既存 DB は部分更新されません。commit: 8f838c2。検証: crawler unit 353/353、crawler E2E 74/74 (skip 0)、typecheck/lint/format/knip。

File: apps/crawler/tests/e2e/cash-flow-history.test.ts [hunk 9]
```diff
@@ -58,10 +61,32 @@ describe("scrapeCashFlowHistory", () => {
...
+
+      const currentMonth = getJstYearMonthKey();
+      const [year, month] = currentMonth.split("-").map(Number);
+      const previousMonth = new Date(year!, month! - 2, 1);
+      const rangeHeader = `${previousMonth.getFullYear()}/${previousMonth.getMonth() + 1}/26 - ${year}/${month}/25`;
+      await page.evaluate((header) => {
+        document.querySelector("a[href*='/cf/csv']")?.remove();
+        const title = document.querySelector(".fc-header-title h2");
+        if (!title) throw new Error("Cash flow month header is unavailable");
+        title.textContent = header;
```
[the remark this answers is not in the document]
Comment (hiroppy) [reply apps/crawler/tests/e2e/cash-flow-history.test.ts]: 対応しました。DOM の CSV link/header 書き換えを削除し、認証済み実サービスで未来の空会計期間へ next navigation した後、production の ensureCurrentCashFlowView（getCashFlow からも利用）で Today navigation を実行する E2E に変更しました。ヘッダー復帰、テーブル表示、実 /cf/fetch の完了を検証しています。synthetic な締め日境界は DOM-independent unit test に移しました。commit: 8f838c2。検証: 対象 E2E 3/3、全 crawler E2E 74/74 (skip 0)。

File: apps/crawler/src/scrapers/cash-flow.ts [hunk 10]
```diff
@@ -13,67 +48,100 @@ export async function getCashFlow(page: Page): Promise<CashFlowSummary> {
   // テーブルが表示されるまで待機
   await page.locator("#cf-detail-table").waitFor({ state: "visible", timeout: 10000 });
 
-  // Click "today" button to ensure we're viewing the current month
+  const currentMonth = getJstYearMonthKey();
```
[the remark this answers is not in the document]
Comment (hiroppy) [reply apps/crawler/src/scrapers/cash-flow.ts]: 対応しました。calendar month の一致ではなく、表示中の period_start <= JST today <= period_end で current accounting period を判定するよう変更しました。これにより 8/26 に 8/26–9/25 が 2026-09 とラベルされても Today を再クリックしません。締め日境界の unit test と、実サービスの空期間から Today へ戻る E2E を追加しています。commit: 8f838c2。検証: crawler unit 353/353、全 crawler E2E 74/74 (skip 0)。

File: apps/crawler/src/crawler-phases.test.ts [hunk 11]
```diff
@@ -278,9 +279,67 @@ describe("runSavePhase", () => {
...
+    vi.mocked(hasCashFlowPeriod).mockResolvedValue(false);
+    vi.mocked(scrapeCashFlowHistory).mockResolvedValue([]);
+
+    await runCashFlowHistoryPhase({} as never, {} as never, { isHistoryMode: true });
+
+    expect(scrapeCashFlowHistory).toHaveBeenCalledWith(
+      {},
+      getHistoryMaxMonths(new Date()),
+      expect.any(Object),
+    );
```
[the remark this answers is not in the document]
Comment (hiroppy) [reply apps/crawler/src/crawler-phases.test.ts]: 対応しました。テスト内で 2026-08-31T14:59:59Z に clock を固定し、runCashFlowHistoryPhase と getHistoryMaxMonths の期待値を同じ instant から算出するよう変更しました。finally で real timers に復元しています。commit: 8b1588a。検証: crawler unit 353/353、全 crawler E2E 74/74 (skip 0)。

File: apps/crawler/src/scrapers/cash-flow.ts [hunk 12]
```diff
@@ -1,10 +1,85 @@
...
+  if (!displayedState) {
+    throw new Error("Could not determine the displayed cash flow month");
+  }
+
+  if (isCurrentAccountingPeriod(displayedState)) return displayedState;
+
+  const todayButton = page.locator(".fc-button-today").first();
+  if (!(await todayButton.isVisible())) {
+    throw new Error(`Could not navigate cash flow from ${displayedState.month} to today`);
+  }
```
[the remark this answers is not in the document]
Comment (hiroppy) [reply apps/crawler/src/scrapers/cash-flow.ts:55]: 対応しました。.fc-button-today に waitFor({ state: "visible", timeout: 3000 }) を入れてから isVisible() を確認し、timeout 後も従来の明示的な error path/message を維持しています。commit: 8b1588a。検証: crawler unit 353/353、実サービスを使う全 crawler E2E 74/74 (skip 0)。

File: packages/db/src/repositories/transactions.ts [hunk 13]
```diff
@@ -183,28 +240,43 @@ export async function replaceTransactionsForMonth(
...
+  const currentYear = parseInt(month.slice(0, 4), 10);
+  const replacementRange = resolveTransactionDateRange(month, dateRange);
+  const toExclusive = getExclusiveRangeEnd(replacementRange.to);
+  const records = items.map((item) => prepareTransactionData(item, accountIdMap, currentYear));
 
-  if (validItems.length === 0) {
-    return 0;
+  if (items.length === 0 && !isComplete) {
+    throw new Error("Cannot replace an empty cash flow period without completeness proof");
+  }
```
[the remark this answers is not in the document]
Comment (hiroppy) [reply packages/db/src/repositories/transactions.ts]: 対応しました。items の件数に関係なく isComplete=false を delete/write 前に拒否するよう変更しました。非空の部分入力を渡し、例外後も既存 transaction IDs が tx1/tx2 のまま残る回帰テストも追加しています。commit: 8b1588a。検証: DB tests 396/396、crawler unit 353/353、全 crawler E2E 74/74 (skip 0)、typecheck/lint/format/knip。

File: apps/crawler/src/crawler-phases.ts [hunk 14]
```diff
@@ -257,28 +257,31 @@ export async function runCashFlowHistoryPhase(
...
+    for (let i = 2; i < maxMonths; i++) {
+      const month = getHistoryMonth(now, i);
+      if (!(await hasCashFlowPeriod(db, month))) {
```
[the remark this answers is not in the document]
Comment (hiroppy) [reply apps/crawler/src/crawler-phases.ts:282]: 対応しました。runScrapePhase で取得済みの current cash-flow accounting month を runCashFlowHistoryPhase へ渡し、coverage key と最大取得数の両方をその月から算出するよう変更しました。締め日後の 2026-09 起点で missing 2026-07 を3期間目として検出するテストと、2025-01まで21期間になる境界テストを追加しています。commit: e70c66a。検証: crawler unit 359/359、全 crawler E2E 74/74 (skip 0)。

File: apps/crawler/src/scrapers/cash-flow-history.ts [hunk 15]
```diff
@@ -215,26 +379,38 @@ export async function extractCashFlowFromPage(page: Page): Promise<CashFlowSumma
...
-    if (item) items.push(item);
+    items.push(await parseDetailRow(detailRows.nth(i), year, { periodStart, periodEnd }));
   }
 
-  return { month, totalIncome, totalExpense, balance, items };
+  return {
+    month,
+    periodStart,
+    periodEnd,
+    isComplete: true,
```
[the remark this answers is not in the document]
Comment (hiroppy) [reply apps/crawler/src/scrapers/cash-flow-history.ts:420]: 対応しました。summary 3セルを対応する金額形式として検証し、空文字やロード中表示を拒否します。明細0件は「CSV linkなし」かつ「収入・支出・残高が全て0」の実サービスで確認した empty-state の場合だけ complete とし、それ以外は保存前に失敗させます。getCashFlow と history extraction の両方へ適用し、空summary、CSVあり0件、非0summaryの回帰テストを追加しました。commit: e70c66a。検証: crawler unit 359/359、全 crawler E2E 74/74 (skip 0)、typecheck/lint/format/knip。

**Comment by hiroppy** [comment]
CodeRabbit の review body 内に残っていた指摘を再確認し、commit 1a3334b で対応しました。

- `runSavePhase.historyMonths` の型へ `dateRange` を追加
- バッチ内の実効日付範囲を全件事前検証し、重複時は delete/write 前に拒否（既存データ保持の回帰テスト追加）
- `saveScrapedDataBatch` の本番経路にも同じ重複検証を適用
- 月跨ぎ・年跨ぎヘッダーの会計月、期間境界、固定時計をテスト
- 不正 dateRange と group selector 検証不一致の回帰テストを追加

月跨ぎヘッダーの終端月パースと重複していた browser-context month parser の除去は現行コードですでに対応済みでした。`waitForFunction` は現行実装では AJAX 適用待ちの1回だけなので、該当指摘は stale と判断しています。

検証: crawler unit 361/361、DB 398/398、crawler E2E 74/74 (skip 0)、typecheck 8/8、lint、format、knip。最新コードで隔離履歴クロール20/20期間も成功し、本番DBは読み取り専用で照合済みです。

**Comment by hiroppy** [comment]
ブランチ全体へ `$simplify` を適用し、commit 0939187 で push しました。

- 現在月／履歴月の cash-flow 抽出を一本化
- 単件保存／期間置換の取引レコード変換を一本化
- 期間置換データ型を共通化
- 重複期間検査を単純なソート＋走査へ整理
- 挙動を維持したまま 5 files, +47/-184（純減137行）

検証: crawler unit 361/361、DB 398/398、crawler E2E 74/74 (skip 0)、typecheck 8/8、lint、format、knip。最新コードで隔離DBへの履歴クロール20/20期間、本番DBとの共通取引・振替・実アプリ月次集計照合も成功しています。GitHub CIは全check green、未解決review threadは0です。
